### PR TITLE
C#: Re-factor tainttracking configurations to use the new API.

### PR DESCRIPTION
**Pull request by michaelnebel**

### CODE REVIEW

**Comment by michaelnebel** [comment]
@aschackmull : In this PR we re-write 7 data dataflow queries to use the new API for C# (maybe out of around 40 in total). According to DCA the cache size increases by 30%. Is this expected/acceptable?

**Comment by aschackmull** [comment]
> @aschackmull : In this PR we re-write 7 data dataflow queries to use the new API for C# (maybe out of around 40 in total). According to DCA the cache size increases by 30%. Is this expected/acceptable?

I think that's been the general trend, yes.

File: csharp/ql/lib/semmle/code/csharp/security/cryptography/EncryptionKeyDataFlowQuery.qll [hunk 1]
```diff
@@ -66,15 +66,18 @@ class SymmetricEncryptionCreateDecryptorSink extends SymmetricEncryptionKeySink
 /**
  * Symmetric Key Data Flow configuration.
  */
-class SymmetricKeyTaintTrackingConfiguration extends TaintTracking::Configuration {
```
Comment (aschackmull) [inline csharp/ql/lib/semmle/code/csharp/security/cryptography/EncryptionKeyDataFlowQuery.qll:69]: You'll need to keep the old config around and deprecate it.

Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/cryptography/EncryptionKeyDataFlowQuery.qll:69]: Oh right!

File: csharp/ql/lib/semmle/code/csharp/security/cryptography/HardcodedSymmetricEncryptionKey.qll [hunk 2]
```diff
@@ -62,6 +62,8 @@ module HardcodedSymmetricEncryptionKey {
   }
 
   /**
+   * DEPRECATED: Use `HardCodedSymmetricEncryption` instead.
+   *
    * A taint-tracking configuration for uncontrolled data in path expression vulnerabilities.
    */
   class TaintTrackingConfiguration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/security/cryptography/HardcodedSymmetricEncryptionKey.qll:69]: missing `deprecated`.

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/CodeInjectionQuery.qll [hunk 3]
```diff
@@ -25,6 +25,8 @@ abstract class Sink extends DataFlow::ExprNode { }
 abstract class Sanitizer extends DataFlow::ExprNode { }
 
 /**
+ * DEPRECATED: Use `CodeInjection` instead.
+ *
  * A taint-tracking configuration for user input treated as code vulnerabilities.
  */
 class TaintTrackingConfiguration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/CodeInjectionQuery.qll:32]: `deprecated`

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/CommandInjectionQuery.qll [hunk 4]
```diff
@@ -23,6 +23,8 @@ abstract class Sink extends DataFlow::ExprNode { }
 abstract class Sanitizer extends DataFlow::ExprNode { }
 
 /**
+ * DEPRECATED: Use `CommandInjection` instead.
+ *
  * A taint-tracking configuration for command injection vulnerabilities.
  */
 class TaintTrackingConfiguration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/CommandInjectionQuery.qll:30]: `deprecated`

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/ConditionalBypassQuery.qll [hunk 5]
```diff
@@ -30,6 +30,8 @@ abstract class Sink extends DataFlow::ExprNode {
 abstract class Sanitizer extends DataFlow::ExprNode { }
 
 /**
+ * DEPRECATED: Use `ConditionalBypass` instead.
+ *
  * A taint-tracking configuration for user-controlled bypass of sensitive method.
  */
 class Configuration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/ConditionalBypassQuery.qll:37]: `deprecated`

File: csharp/ql/lib/semmle/code/csharp/security/dataflow/ExposureOfPrivateInformationQuery.qll [hunk 6]
```diff
@@ -23,6 +23,8 @@ abstract class Sink extends DataFlow::ExprNode { }
 abstract class Sanitizer extends DataFlow::ExprNode { }
 
 /**
+ * DEPRECATED: Use `ExposureOfPrivateInformation` instead.
+ *
  * A taint-tracking configuration for private information flowing unencrypted to an external location.
  */
 class TaintTrackingConfiguration extends TaintTracking::Configuration {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/security/dataflow/ExposureOfPrivateInformationQuery.qll:30]: `deprecated`

File: csharp/ql/src/Security Features/CWE-078/StoredCommandInjection.ql [hunk 7]
```diff
@@ -16,13 +16,19 @@
...
+import StoredCommandInjection::PathGraph
 
-class StoredTaintTrackingConfiguration extends TaintTrackingConfiguration {
-  override predicate isSource(DataFlow::Node source) { source instanceof StoredFlowSource }
+module StoredCommandInjectionConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof StoredFlowSource }
+
+  predicate isSink = CommandInjectionConfig::isSink/1;
+
+  predicate isBarrier = CommandInjectionConfig::isBarrier/1;
```
Comment (hvitved) [inline csharp/ql/src/Security Features/CWE-078/StoredCommandInjection.ql:26]: Alternatively replace with `import CommandInjectionConfig`.

**Review by hvitved — CHANGES_REQUESTED** [review]
LGTM, just missing a couple of deprecations.

File: csharp/ql/lib/semmle/code/csharp/security/cryptography/HardcodedSymmetricEncryptionKey.qll (hunk 2 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/security/cryptography/HardcodedSymmetricEncryptionKey.qll]: Haha, yes!

File: csharp/ql/src/Security Features/CWE-078/StoredCommandInjection.ql (hunk 7 above)
Comment (michaelnebel) [reply csharp/ql/src/Security Features/CWE-078/StoredCommandInjection.ql:26]: That is not possible.

**Pull request merged**